File: app/src/test/java/com/nytimes/android/external/playbillingtester/bundle/ConsumePurchaseResponseTest.java
package com.nytimes.android.external.playbillingtester.bundle;

import android.os.Bundle;

import com.google.common.collect.ImmutableList;
import com.nytimes.android.external.playbillingtester.APIOverrides;
import com.nytimes.android.external.playbillingtester.BillingServiceStubImpl;
import com.nytimes.android.external.playbillingtester.Purchases;
import com.nytimes.android.external.playbillingtesterlib.GoogleUtil;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyListOf;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ConsumePurchaseResponseTest {

    private static final int API_VERSION = 3;
    private static final String PACKAGE_NAME = "com.my.package";

    @Mock
    private APIOverrides apiOverrides;

    @Mock
    private Purchases purchases;

    @Mock
    private Purchases.PurchasesLists inAppPurchasesLists;

    @Mock
    private Purchases.PurchasesLists subscriptionsPurchasesLists;

    @Mock
    private BuyIntentBundleBuilder buyIntentBundleBuilder;

    @Mock
    private SkuDetailsBundleBuilder skuDetailsBundleBuilder;

    @Mock
    private PurchasesBundleBuilder purchasesBundleBuilder;

    @Mock
    private BuyIntentToReplaceSkusBundleBuilder buyIntentToReplaceSkusBundleBuilder;

    private BillingServiceStubImpl testObject;

    private final Bundle expected = new Bundle();
    private final String type = GoogleUtil.BILLING_TYPE_IAP;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        when(skuDetailsBundleBuilder.newBuilder()).thenReturn(skuDetailsBundleBuilder);
        when(skuDetailsBundleBuilder.skus(anyListOf(String.class), anyString())).thenReturn(skuDetailsBundleBuilder);
        when(skuDetailsBundleBuilder.build()).thenReturn(expected);
        when(buyIntentBundleBuilder.newBuilder()).thenReturn(buyIntentBundleBuilder);
        when(buyIntentBundleBuilder.developerPayload(anyString())).thenReturn(buyIntentBundleBuilder);
        when(buyIntentBundleBuilder.packageName(anyString())).thenReturn(buyIntentBundleBuilder);
        when(buyIntentBundleBuilder.sku(anyString())).thenReturn(buyIntentBundleBuilder);
        when(buyIntentBundleBuilder.type(anyString())).thenReturn(buyIntentBundleBuilder);
        when(buyIntentBundleBuilder.build()).thenReturn(expected);
        when(purchasesBundleBuilder.newBuilder()).thenReturn(purchasesBundleBuilder);
        when(purchasesBundleBuilder.type(anyString())).thenReturn(purchasesBundleBuilder);
        when(purchasesBundleBuilder.continuationToken(anyString())).thenReturn(purchasesBundleBuilder);
        when(purchasesBundleBuilder.build()).thenReturn(expected);
        ConsumePurchaseResponse consumePurchaseResponse = new ConsumePurchaseResponse(apiOverrides, purchases);
        testObject = new BillingServiceStubImpl(apiOverrides, buyIntentBundleBuilder, skuDetailsBundleBuilder,
                purchasesBundleBuilder, consumePurchaseResponse, buyIntentToReplaceSkusBundleBuilder);
    }

    @Test
    public void testConsumePurchaseIAP() {
        final List<String> testPurchases = ImmutableList.of("purchase1", "purchase2");
        final List<String> subscriptions = ImmutableList.of("subscription1", "subscription2");
        Bundle getPurchasesBundle = new Bundle();
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, new ArrayList<>(testPurchases));
        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null)).thenReturn(inAppPurchasesLists);
        when(inAppPurchasesLists.purchaseDataList()).thenReturn(testPurchases);
        when(purchases.removePurchase("purchase1", GoogleUtil.BILLING_TYPE_IAP)).thenReturn(true);

        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists);
        when(subscriptionsPurchasesLists.purchaseDataList()).thenReturn(subscriptions);

        when(purchasesBundleBuilder.build()).thenReturn(getPurchasesBundle);

        Bundle stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token");
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK);
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases);
        assertThat(stored).isEqualTo(getPurchasesBundle);

        String purchaseToken = testPurchases.get(0);
        int result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken);
        assertThat(result).isEqualTo(GoogleUtil.RESULT_OK);
    }

    @Test
    public void testConsumePurchaseSubscription() {
        final List<String> testPurchases = ImmutableList.of("purchase1", "purchase2");
        final List<String> subscriptions = ImmutableList.of("subscription1", "subscription2");
        Bundle getPurchasesBundle = new Bundle();
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, new ArrayList<>(testPurchases));
        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null))
                .thenReturn(inAppPurchasesLists);
        when(inAppPurchasesLists.purchaseDataList()).thenReturn(testPurchases);
        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists);
        when(subscriptionsPurchasesLists.purchaseDataList())
                .thenReturn(subscriptions);
        when(purchasesBundleBuilder.build())
                .thenReturn(getPurchasesBundle);

        Bundle stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token");
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK);
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases);
        assertThat(stored).isEqualTo(getPurchasesBundle);

        String purchaseToken = subscriptions.get(0);

        int result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken);
        assertThat(result).isEqualTo(GoogleUtil.RESULT_ERROR);
    }

    @Test
    public void testConsumePurchaseNotOwned() {
        final List<String> testPurchases = ImmutableList.of("purchase1", "purchase2");
        final List<String> subscriptions = ImmutableList.of("subscription1", "subscription2");
        Bundle getPurchasesBundle = new Bundle();
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, new ArrayList<>(testPurchases));
        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null))
                .thenReturn(inAppPurchasesLists);
        when(inAppPurchasesLists.purchaseDataList()).thenReturn(testPurchases);
        when(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists);
        when(subscriptionsPurchasesLists.purchaseDataList())
                .thenReturn(subscriptions);
        when(purchasesBundleBuilder.build())
                .thenReturn(getPurchasesBundle);

        Bundle stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token");
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK);
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases);
        assertThat(stored).isEqualTo(getPurchasesBundle);

        String purchaseToken = "Not owned product";

        int result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken);
        assertThat(result).isEqualTo(GoogleUtil.RESULT_ITEM_NOT_OWNED);
    }

}

File: app/src/test/java/com/nytimes/android/external/register/bundle/ConsumePurchaseResponseTest.kt
package com.nytimes.android.external.register.bundle

import android.os.Bundle
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.common.collect.ImmutableList
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.nytimes.android.external.register.APIOverrides
import com.nytimes.android.external.register.BillingServiceStubImpl
import com.nytimes.android.external.register.Purchases
import com.nytimes.android.external.register.PurchasesLists
import com.nytimes.android.external.registerlib.GoogleUtil
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers.anyListOf
import org.mockito.Matchers.anyString
import java.util.*

@RunWith(AndroidJUnit4::class)
class ConsumePurchaseResponseTest {

    private val apiOverrides: APIOverrides = mock()

    private val purchases: Purchases = mock()

    private val inAppPurchasesLists: PurchasesLists = mock()

    private val subscriptionsPurchasesLists: PurchasesLists = mock()

    private val buyIntentBundleBuilder: BuyIntentBundleBuilder = mock()

    private val skuDetailsBundleBuilder: SkuDetailsBundleBuilder = mock()

    private val purchasesBundleBuilder: PurchasesBundleBuilder = mock()

    private val buyIntentToReplaceSkusBundleBuilder: BuyIntentToReplaceSkusBundleBuilder = mock()

    private lateinit var testObject: BillingServiceStubImpl

    private val expected = Bundle()

    private val type = GoogleUtil.BILLING_TYPE_IAP

    @Before
    fun setUp() {
        whenever(skuDetailsBundleBuilder.newBuilder()).thenReturn(skuDetailsBundleBuilder)
        whenever(skuDetailsBundleBuilder.skus(anyListOf(String::class.java), anyString())).thenReturn(skuDetailsBundleBuilder)
        whenever(skuDetailsBundleBuilder.build()).thenReturn(expected)
        whenever(buyIntentBundleBuilder.newBuilder()).thenReturn(buyIntentBundleBuilder)
        whenever(buyIntentBundleBuilder.developerPayload(anyString())).thenReturn(buyIntentBundleBuilder)
        whenever(buyIntentBundleBuilder.packageName(anyString())).thenReturn(buyIntentBundleBuilder)
        whenever(buyIntentBundleBuilder.sku(anyString())).thenReturn(buyIntentBundleBuilder)
        whenever(buyIntentBundleBuilder.type(anyString())).thenReturn(buyIntentBundleBuilder)
        whenever(buyIntentBundleBuilder.build()).thenReturn(expected)
        whenever(purchasesBundleBuilder.newBuilder()).thenReturn(purchasesBundleBuilder)
        whenever(purchasesBundleBuilder.type(anyString())).thenReturn(purchasesBundleBuilder)
        whenever(purchasesBundleBuilder.continuationToken(anyString())).thenReturn(purchasesBundleBuilder)
        whenever(purchasesBundleBuilder.build()).thenReturn(expected)
        val consumePurchaseResponse = ConsumePurchaseResponse(apiOverrides, purchases)
        testObject = BillingServiceStubImpl(apiOverrides, buyIntentBundleBuilder, skuDetailsBundleBuilder,
                purchasesBundleBuilder, consumePurchaseResponse, buyIntentToReplaceSkusBundleBuilder)
    }

    @Test
    fun testConsumePurchaseIAP() {
        val testPurchases = mutableListOf("purchase1", "purchase2")
        val subscriptions = mutableListOf("subscription1", "subscription2")
        val getPurchasesBundle = Bundle()
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, ArrayList(testPurchases))
        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null)).thenReturn(inAppPurchasesLists)
        whenever(inAppPurchasesLists.purchaseDataList).thenReturn(testPurchases)
        whenever(purchases.removePurchase("purchase1", GoogleUtil.BILLING_TYPE_IAP)).thenReturn(true)

        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists)
        whenever(subscriptionsPurchasesLists.purchaseDataList).thenReturn(subscriptions)

        whenever(purchasesBundleBuilder.build()).thenReturn(getPurchasesBundle)

        val stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token")
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK)
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases)
        assertThat(stored).isEqualTo(getPurchasesBundle)

        val purchaseToken = testPurchases[0]
        val result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken)
        assertThat(result).isEqualTo(GoogleUtil.RESULT_OK)
    }

    @Test
    fun testConsumePurchaseSubscription() {
        val testPurchases = ImmutableList.of("purchase1", "purchase2")
        val subscriptions = ImmutableList.of("subscription1", "subscription2")
        val getPurchasesBundle = Bundle()
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, ArrayList(testPurchases))
        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null))
                .thenReturn(inAppPurchasesLists)
        whenever(inAppPurchasesLists.purchaseDataList).thenReturn(testPurchases)
        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists)
        whenever(subscriptionsPurchasesLists.purchaseDataList)
                .thenReturn(subscriptions)
        whenever(purchasesBundleBuilder.build())
                .thenReturn(getPurchasesBundle)

        val stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token")
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK)
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases)
        assertThat(stored).isEqualTo(getPurchasesBundle)

        val purchaseToken = subscriptions[0]

        val result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken)
        assertThat(result).isEqualTo(GoogleUtil.RESULT_ERROR)
    }

    @Test
    fun testConsumePurchaseNotOwned() {
        val testPurchases = ImmutableList.of("purchase1", "purchase2")
        val subscriptions = ImmutableList.of("subscription1", "subscription2")
        val getPurchasesBundle = Bundle()
        getPurchasesBundle.putStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST, ArrayList(testPurchases))
        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_IAP, null))
                .thenReturn(inAppPurchasesLists)
        whenever(inAppPurchasesLists.purchaseDataList).thenReturn(testPurchases)
        whenever(purchases.getPurchasesLists(GoogleUtil.BILLING_TYPE_SUBSCRIPTION, null))
                .thenReturn(subscriptionsPurchasesLists)
        whenever(subscriptionsPurchasesLists.purchaseDataList)
                .thenReturn(subscriptions)
        whenever(purchasesBundleBuilder.build())
                .thenReturn(getPurchasesBundle)

        val stored = testObject.getPurchases(API_VERSION, PACKAGE_NAME, type, "token")
        assertThat(stored.getInt(GoogleUtil.RESPONSE_CODE))
                .isEqualTo(GoogleUtil.RESULT_OK)
        assertThat(stored.getStringArrayList(GoogleUtil.INAPP_PURCHASE_DATA_LIST))
                .isEqualTo(testPurchases)
        assertThat(stored).isEqualTo(getPurchasesBundle)

        val purchaseToken = "Not owned product"

        val result = testObject.consumePurchase(API_VERSION, PACKAGE_NAME, purchaseToken)
        assertThat(result).isEqualTo(GoogleUtil.RESULT_ITEM_NOT_OWNED)
    }

    companion object {
        private const val API_VERSION = 3
        private const val PACKAGE_NAME = "com.my.package"
    }

}
